### DISCUSSION TITLE
Exception raised on "transfer-encoding: identity"

### DISCUSSION THREAD

**Discussion by eshen1991** [mitmproxy/mitmproxy#7360] [Q&A] [answered]
I am proxying [GCS json API](https://cloud.google.com/storage/docs/json_api) through mitmproxy. The following is a GCS API exported as curl: 
```
curl -H 'User-Agent: TSL' -H 'Accept: */*' -H 'Authorization: Bearer token' -H 'X-Upload-Content-Length: 22' -H 'Transfer-Encoding: identity' -H 'Content-Type: application/x-www-form-urlencoded' -H 'Expect: 100-continue' -X POST 'https://www.googleapis.com/upload/storage/v1/b/eshen-gcs-proxy-2/o?uploadType=resumable&name=1732773308203%2Ffunc-test-ff11df61-adb6-43e0-a1e9-775169035db8'
```
The request has a Transfer-Encoding header and the value is "identity", for which mitmproxy [raises exception](https://github.com/mitmproxy/mitmproxy/blob/main/mitmproxy/net/http/http1/read.py#L113-L118). 

I don't have control on the application that sends the request and I can't change the request headers. So I am just wondering if there is any workaround for mitmproxy not raising errors on  'Transfer-Encoding: identity'

**Comment by mhils** [comment, marked as answer]
Technically this is invalid HTTP, but https://github.com/mitmproxy/mitmproxy/pull/7361 should make it work if the `validate_inbound_headers` option is disabled. 😃

**Reply by eshen1991** [reply]
@mhils Thanks so much for such a quick response and fix(not to mention it's around Thanksgiving). Really appreciate it!  Now the request bypassed the error around te.  The same request is supposed to get a [session URI](https://cloud.google.com/storage/docs/resumable-uploads#session-uris) in the response. However, the request got "[server closed connection](https://github.com/mitmproxy/mitmproxy/blob/18eb29c7364bb5dafab1af3e4259e70b9a5fbfc3/mitmproxy/proxy/layers/http/_http1.py#L440-L445)" error. I didn't see any detailed errors in the mitmproxy log.  I have several questions:

- Is there a way to turn on the log level for mitmproxy so I can potentially see more detailed errors regarding "server closed connection" error?
- Is there a way to set a longer http timeout with a server that mitmproxy connects to? 
- In general, how to troubleshoot issues like "server closed connection" within mitmproxy? 

Here is how I started mitmproxy: 
```
mitmweb --set confdir=/Users/ericshen/mitmproxy --set validate_inbound_headers=false -s mitmproxy-gcs.py -w capture.mitm
```
By the way, the same code(request) works without going through mitmproxy. Pleas let me know if you'd need more info.

**Reply by mhils** [reply]
> Is there a way to turn on the log level for mitmproxy so I can potentially see more detailed errors regarding "server closed connection" error?

`-vvv --set proxy_debug` will give you very verbose logs.

> Is there a way to set a longer http timeout with a server that mitmproxy connects to?

No, but that is unlikely to be the issue here. We currently default to 60s without any activity IIRC, which should be plenty.

> In general, how to troubleshoot issues like "server closed connection" within mitmproxy?

Either with the command listed above, or with https://docs.mitmproxy.org/stable/howto-wireshark-tls/.


Based on what you describe, my hunch would be that `--set stream_large_bodies=1` may make things work. Please let me know, that'd be useful feedback to adjust our heuristic. :)

**Reply by eshen1991** [reply]
@mhils Again appreciate your quick response and thanks for the troubleshooting tips! I tried your suggestions and observed the following:

- The request (curl command posted earlier) works with http/2.0. I get the correct response with and without the mitmproxy. 
- The request fails with http/1.1(the app sends the request in http/1.1)
   - With proxy 
       - My custom addon for request didn't get called at all. 
       - I get the "server closed connection" error
       - I also noticed that the proxy added a header "Expect: 100-continue"
   - Without the proxy, I get the following error directly from GCS:  
      -  <p>POST requests require a <code>Content-length</code> header.  <ins>That’s all we know.</ins>

My follow-up questions:

- Why the proxy adds "Expect: 100-continue" for http/1.1
-  I expect the "require a Content-length header" error I got from sending the request directly to GCS would have been returned to the mitmproxy when the proxy was used. Does the proxy handle http/1.1 in some special ways? 
- If I were to change the header or http version of the request before it gets processed by the native addons, how would I do this. Or is this even possible as all native addons always get triggered first?   

Please let me know if I'd need to create a new discussion topic if this thread is getting too long.

**Reply by mhils** [reply]
> Without the proxy, I get the following error directly from GCS:

To confirm, without the proxy it does not work either?

> I also noticed that the proxy added a header "Expect: 100-continue"

I would be highly surprised if mitmproxy added that header. Major claims require major evidence. :)

> My custom addon for request didn't get called at all.

If you define a `requestheaders` hook, is that being called? If yes, can you try setting `flow.request.stream = True` in there and see if that changes things?

**Reply by eshen1991** [reply]
> My custom addon for request didn't get called at all.

The following is how I registered my addon.  _log_request would print out the request. I didn't see any log info from _log_request. 
```
def request(flow: http.HTTPFlow) -> None:
  """Intercepts and potentially modifies HTTP requests.

  This function specifically targets GCS upload requests, attempting to
  encrypt the content before it's uploaded.

  Args:
      flow: The HTTP flow object representing the request/response exchange.
  """
  try:
    _log_request(flow, "###### request begin #####")


def start():
  """Entry point for starting the mitmproxy addon."""
  from mitmproxy import ctx
  ctx.master.addons.add(request)
```

> If you define a requestheaders hook, is that being called? If yes, can you try setting flow.request.stream = True in there and see if that changes things?

Given the code snippet above, can you elaborate a little more on what you suggested. Maybe point me to some examples.

**Reply by eshen1991** [reply]
> > Without the proxy, I get the following error directly from GCS:

> To confirm, without the proxy it does not work either?

Yes, without proxy, using curl http/1.1 returns 411 Length Required.  The following is my curl command: 
```
curl --http1.1 -v -H 'User-Agent: TSL' --compressed -H 'Accept: */*' -H 'Connection: keep-alive' -H 'Authorization: Bearer token' -H 'X-Upload-Content-Length: 22' -H 'Transfer-Encoding: identity' -H 'Content-Type: application/x-www-form-urlencoded' -H 'x-eshen: something'  -X POST 'https://www.googleapis.com/upload/storage/v1/b/eshen-gcs-proxy-2/o?uploadType=resumable&name=1733165372239%2Ffunc-test-eef5974e-adba-4457-a57d-257af214dd9b'
```

**Reply by mhils** [reply]
> Given the code snippet above, can you elaborate a little more on what you suggested. Maybe point me to some examples.

request -> requestheaders (https://docs.mitmproxy.org/stable/api/events.html#available-hooks)

> Yes, without proxy, using curl http/1.1 returns 411 Length Required. The following is my curl command:

This makes me think that we shouldn't be supporting this at all - I thought this was at least tolerated by some servers in practice. But now we have a client that sends something that looks like HTTP but that is not valid HTTP, and no servers accepting it. I don't think mitmproxy should necessarily accept it either.

**Reply by eshen1991** [reply]
Let me try the requestheaders to remove the TE header. 

Also just FYI, regarding "Expect: 100-continue", I saw that from the web UI (see the attached file). The header was added when the request was being processed and was gone when the request failed. 
[mitmproxy-web-screen-capture.pdf](https://github.com/user-attachments/files/17985092/mitmproxy-web-screen-capture.pdf)

**Reply by mhils** [reply]
Looking at https://github.com/search?q=repo%3Amitmproxy%2Fmitmproxy%20100-continue&type=code, I would be highly surprised if mitmproxy added that header somehow. :)

**Reply by eshen1991** [reply]
I think 100-continue is added by [libcurl](https://github.com/curl/curl/blob/90d937def998f7499d1ab307b70b35dd7e4991ae/docs/libcurl/opts/CURLOPT_HTTPPOST.md?plain=1#L43) for http1.1
